pkg/devfile/types.go

```diff
@@ -8,31 +8,31 @@ import (
 	corev1 "k8s.io/api/core/v1"
 )
 
-// devfileForm is the needed data to send to the devfile library
-type devfileForm struct {
+// DevfileForm is the needed data to send to the devfile library
+type DevfileForm struct {
 	Name    string      `json:"name"`
-	Git     gitData     `json:"git"`
-	Devfile devfileData `json:"devfile"`
+	Git     GitData     `json:"git"`
+	Devfile DevfileData `json:"devfile"`
 }
 
-// devfileSamplesForm is the needed data to query the devfile registry for samples
-type devfileSamplesForm struct {
+// DevfileSamplesForm is the needed data to query the devfile registry for samples
+type DevfileSamplesForm struct {
 	Registry string `json:"registry"`
 }
 
-// devfileResources is the constructed response from the devfile library data
-type devfileResources struct {
+// DevfileResources is the constructed response from the devfile library data
+type DevfileResources struct {
 	ImageStream imagev1.ImageStream `json:"imageStream"`
 	// This can be enhanced to include BuildResource Type that includes all possible types of build objects(eg buildConfig, build, pod, etc.)
 	BuildResource buildv1.BuildConfig `json:"buildResource"`
 	// This can be enhanced to include Deploy Resource Type that includes all possible types of deployment objects(eg deployment, deploymentConfig, helmChart, etc.)
 	DeployResource appsv1.Deployment `json:"deployResource"`
-	Service        corev1.Service    `json:"service"`
-	Route          routev1.Route     `json:"route"`
+	Service        *corev1.Service   `json:"service"`
+	Route          *routev1.Route    `json:"route"`
```

Comment:
This pointers return this json if the Service or route is not defined:

```json
{
  /* ... */
  "service": null,
  "route": null
}
```

You can omit this null values, this would fix the crash you mentioned here https://github.com/openshift/console/pull/12000#issuecomment-1282804735. But it still need some frontend changes of PR #12168. So its up to you if you want apply this small change or not:

```suggestion
	Service        *corev1.Service   `json:"service,omitempty"`
	Route          *routev1.Route    `json:"route,omitempty"`
```

Reply:
@jerolimov oops, I have committed this. if you can, please lgtm it again, thanks!